doc/CONTRIBUTING.md

```diff
@@ -70,6 +70,85 @@ or reported for next major release if BC is not possible.
 
 If you are not sure of what to do, don't hesitate to open an issue about your PR project.
 
+#### Composing the PR
+
+Nobody is perfect, and submitting perfect changes that don't need any tweaks is
+quite a rare instance. After a few iterations, your PR, which is a list of
+commits, could look like this:
+
+- replace tabs with spaces (who did this?)
+- add flux capacitor feature
+- add missing documentation (silly me)
+- add missing tests (I love TDD, but…)
+- fix typo in documentation
+- set default value as suggested
+
+Ideally, we should be able to revert any commit in your PR and not have to
+revert the others too. For instance, if we decide to revert your feature
+because it turns out to cause more problems than it solves, we should not have
+to revert the commit with the feature itself, then the commit with the
+documentation, then the commit with the tests, etc…
+
+Another thing to take into consideration is that in some hopefully rare
+instances, you will need to keep some commits separate, because they do not
+really relate to what you are doing, but the "other" thing you fixed, while
+unrelated, somehow prevents you from doing your PR, or is a minor thing you
+feel does not warrant a PR by itself, especially if this PR is conflicting with
```

Comment:
> Big :-1: on that, you know I'll will. :-)

Totally knew you would, but did it nonetheless because it is easier to discuss on a concrete proposition, and I knew it would make you voice your concerns more clearly. Let me address them.

> Yeah, on rare case it could happens, but it's not a reason to add it on contributing section.

In retrospect, I don't think it is so rare. The opposite, actually. Very very often, before diving in my feature, I have to change something phpcs-fixer cannot fix first.

> BTW, PR with multiple "justified" commit are dangerous because all people are not git masters and could do some.

This is why we have code review. Git masters can review what the git padawans want to merge and tell them.

> Imagine two "justified" commits A and B. The contributor add a "A fix commit" and when he tries to rewrite the history, accidentally fixed up on the B commit. How to fix that? AFAIK, beside redo all the commits, I don't think you can.

Then you need to learn about `git reflog`. Type `git reflog`, recognize the hash of the commit you were on before rebasing (which is pretty obvious when you look at the output of `git reflog`), and `reset` to it. Then you can try your rebase again.

> So I maybe feels insistent, but PR should concern one and only one thing.
Adopting the simple one line rule "If this line does not concern your change goal, does not touch it.".

This is very very wrong IMO. It it were a very clean project, why not but how many times do you come accross a line that makes you think "can I leave this like that, really?" when contributing? The boy-scout rule is better IMO, and we should not prevent it being applied. You're proposing several parallel PRs for that, but I think it will drive people crazy.

Finally, you say that "squash and merge" can make us avoid teaching `git rebase --interactive` to people, and I say they absolutely *need* to know about it. They are developers, and they should be able to produce clean history. And when you get them to try rebase, I'm sure they start using it on their projects, and come back better contributors. Don't fish for them with "squash and merge", try to teach them how to fish first.